src/services/factories/tokenList.ts

```diff
@@ -130,19 +132,35 @@ export function getTokensFactory(
     return new Map(tokenAddressIdPairs)
   }
 
+  async function fetchTcrAddresses(networkId: number): Promise<Set<string>> {
+    return new Set(await tcrApi.getTokens(networkId))
+  }
+
   async function updateTokenDetails(networkId: number, numTokens: number): Promise<void> {
+    // Fetch addresses from TCR
+    const tcrAddressesSet = await retry(() => fetchTcrAddresses(networkId))
+    logDebug(`[tokenListFactory][${networkId}] TCR contains ${tcrAddressesSet.size} addresses`)
+
     // Fetch addresses from contract given numTokens count
     const addressesAndIds = await retry(() => fetchAddressesAndIds(networkId, numTokens))
 
     logDebug(`[tokenListFactory][${networkId}] Token id and address mapping:`)
-    addressesAndIds.forEach((id, address) => logDebug(`[tokenListFactory][${networkId}] ${id} : ${address}`))
+    addressesAndIds.forEach((id, address) => {
+      // Remove addresses that are not on the TCR, if any
+      let removed = false
+      if (tcrAddressesSet.size > 0 && !tcrAddressesSet.has(address)) {
```

Comment:
`tcrAddressesSet.size` don't change between iterations.

I would invert it to look like:
```js
if (tcrAddressesSet.size > 0) {
     addressesAndIds = addressesAndIds.filter(...)
}
```

Reply:
No, but I still want to log all addresses, so a conditional filter just mean duplicated code